rust/src/nfs/nfs4_records.rs

```diff
@@ -829,3 +829,270 @@ pub fn parse_nfs4_response_compound(i: &[u8]) -> IResult<&[u8], Nfs4ResponseComp
+
+        let (_, request) = nfs4_req_close(&buf[4..]).unwrap();
+        match request {
+            Nfs4RequestContent::Close(req_stateid) => {
+                assert_eq!(req_stateid, stateid_buf);
+            }
+            _ => { panic!("Failure"); }
+        }
+    }
+
+    #[test]
+    fn test_nfs4_request_sequenece() {
+    // SEQUENCE op_bytes
+        let buf: &[u8] = &[
+            0x00, 0x00, 0x00, 0x35, /*opcode*/
+            0x00, 0x00, 0x02, 0xd2, 0xe0, 0x14, 0x82, 0x00, /*ssn_id*/
+            0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04, 0x02,
+            0x00, 0x00, 0x00, 0x18, /*_seq_id*/
+            0x00, 0x00, 0x00, 0x00, /*_slot_id*/
+            0x00, 0x00, 0x00, 0x00, /*_high_slot_id*/
+            0x00, 0x00, 0x00, 0x01, /*_catch_this*/
+        ];
+
+        let (_, req_sequence) = nfs4_req_sequence(&buf[4..]).unwrap();
+        match req_sequence {
+            Nfs4RequestContent::Sequence(seq_buf) => {
+                assert_eq!(seq_buf.ssn_id, &buf[4..20]);
+            }
+            _ => { panic!("Failure"); }
+        }
+    }
+
+    #[test]
+    fn test_nfs4_request_lookup() {
+    // LOOKUP op_bytes
+        let buf: &[u8] = &[
+            0x00, 0x00, 0x00, 0x0f, /*opcode*/
+            0x00, 0x00, 0x00, 0x04, 0x76, 0x6f, 0x6c, 0x31, /*fiename: (vol1)*/
+        ];
+
+        let (_, filename_buf) = nfs4_parse_nfsstring(&buf[4..]).unwrap();
+
+        let (_, request) = nfs4_req_lookup(&buf[4..]).unwrap();
+        match request {
+            Nfs4RequestContent::Lookup(req_lookup) => {
+                assert_eq!(req_lookup.filename, filename_buf);
+            }
+            _ => { panic!("Failure!"); }
+        }
+    }
+
+/*
+Testing simple parsers and skipping the record/construct checks
+*/
+    #[test]
+    fn test_nfs4_handle() {
+    // PUTFH op_bytes
+        let buf: &[u8] = &[
+            0x00, 0x00, 0x00, 0x16, /*putfh_opcode*/
+            0x00, 0x00, 0x00, 0x20, /*handle_len*/
+            0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, /*handle*/
+            0x00, 0x00, 0x00, 0x00, 0x84, 0x72, 0x00, 0x00,
+            0x23, 0xa6, 0xc0, 0x12, 0x00, 0xf2, 0xfa, 0x80,
+            0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        ];
+
+        let (_, handle) = nfs4_parse_handle(&buf[4..]).unwrap();
+        assert_eq!(handle.len, 32);
+        assert_eq!(handle.value, &buf[8..]);
+    }
+
+    #[test]
+    fn test_nfs4_attrs() {
+    // GETATTR op_bytes
+        let buf: &[u8] = &[
+            0x00, 0x00, 0x00, 0x09, /*opcode*/
+            0x00, 0x00, 0x00, 0x03, /*attr_cnt*/
+            0x00, 0x00, 0x20, 0x65, /*attr_mask[0]*/
+            0x00, 0x00, 0x00, 0x00, /*attr_mask[1]*/
+            0x00, 0x00, 0x08, 0x00, /*attr_mask[2]*/
+        ];
+
+        let (r, attr) = nfs4_parse_attrbits(&buf[4..]).unwrap();
+        assert_eq!(r.len(), 0);
+        // assert_eq!(attr.attr_mask, 35618163785728);
+        assert_eq!(attr.attr_mask, ((0x00002065 as u64) << 32 | 0 as u64));
```

Comment:
What does the `| 0` do here?

Reply:
`attr_mask` is evaluated as `attr_mask: ((attr_mask1 as u64) << 32) | attr_mask2.unwrap_or(0) as u64`

in this packet `attr_mask[1]` is `0`, I thought to either assert against an expression and substitute the `attr_mask[*]` values as in:
`assert_eq!(attr.attr_mask, ((0x00002065 as u64) << 32 | 0 as u64))` 

or assert against the result like the commented line:
`assert_eq!(attr.attr_mask, 35618163785728)`